### PR TITLE
Converting v3 to key.Destination

### PR DESCRIPTION
**Pull request by alainjobart**
This set of commits adds the new functions, converts all call sites one commit at a time, and then removes the old code.

### CODE REVIEW

**Comment by alainjobart** [comment]
@sougou this is ready for review. I like the way this looks, I think it's somewhat easier to follow.

The KsidOrRange and Ksids structs are now gone, replaced by key.Destination. Also, by using key.DestinationNone{}, it's easier to see where the vindexes Map to nothing, as opposed to creating an empty KsidOrRange.

**Comment by alainjobart** [comment]
@sougou also a couple follow-up questions, for the future after this is committed:

* vindex.IsFunctional() is never used. Same for the old Functional interface. Should we remove it?
* key.Destination.IsUnique() method is also unused. Does it have any value for v3? It was meant to say: this destination will result in at most one outgoing query when used. I don't think v3 cares.

**Comment by sougou** [comment]
If something is unused, we can remove it. I might have created those to follow the original design, which might have made assumptions that they would be useful.

**Comment by alainjobart** [comment]
Also as separate discussion points / future changes:

* VCursor.ExecuteStandalone is just a shortcut for ExecuteMultiShard(..., true, false). Might be just easier to call ExecuteMultiShard.

* VCursor.ExecuteMultiShard uses an array of *querypb.BoundQuery, but we almost always build that array from an array of queries and an array of bind variables map. Maybe it would be easier to take two arrays, []string and []map[string]*querypb.BindVariable.

**Comment by sougou** [comment]
I created ExecuteStandalone mainly for readability. If it's not achieving that purpose, we can delete it :).

That's a thought. You should look at the routingMap structure. Also, the Stream API is different. I thought long and hard about simplifying and unifying the two. Nothing obvious popped. So, I left it alone.

The two slice idea is new. So, if you see simplification through that, go for it!

File: go/vt/vtgate/scatter_conn.go [hunk 1]
```diff
@@ -238,22 +230,17 @@ func (stc *ScatterConn) ExecuteMultiShard(
...
-	shards := make([]string, 0, len(shardQueries))
-	for shard := range shardQueries {
-		shards = append(shards, shard)
-	}
 
-	canCommit := len(shards) == 1 && canAutocommit && session.AutocommitApproval()
+	canCommit := len(rss) == 1 && canAutocommit && session.AutocommitApproval()
 
-	err := stc.multiGoTransaction(
+	err := stc.multiGoTransaction2(
```
Comment (sougou) [inline go/vt/vtgate/scatter_conn.go]: Do we still need the old multigoTransaction?

Comment (alainjobart) [reply go/vt/vtgate/scatter_conn.go]: It's still called by vtgate.Execute, which is still called in a couple places. I'll clean that up next, along with the remaining srvtopo methods that are not used any more. This is a huge PR already, so I'll just merge this now, and work on clean up as a separate one.

**Pull request merged**